### DISCUSSION TITLE
(Fly.io) Significantly delayed database query responses

### DISCUSSION THREAD

**Discussion by ilshm** [pocketbase/pocketbase#4315] [Q&A]
Hello!

I might be missing something, but I've conducted tests across different devices including a phone, PC, and hosting services.

I consistently use `pb.collection('').getOne('')` to retrieve a single record from a collection that contains between 1 to 5 rows.

On hosting, particularly with the free tier of fly.io, I encountered significant delays, approximately 300-500ms, which made it practically unusable.

For local tests, I utilized a PC (equipped with a 10700F processor and 16GB RAM) and a phone (powered by a Snapdragon 865 with 8GB RAM), which should theoretically be more than capable for this task.

Here are my findings:

On the PC (repeating the `pb.collection('').getOne('')` query with a table containing 1-5 records, after several attempts):
```json
{
  "collectionId": "wd3g9giw0wbb5tn",
  "collectionName": "pastes",
  "created": "2024-02-08 16:03:27.354Z",
  "id": "w0eakza33eaubjf",
  "text": "Hello",
  "updated": "2024-02-08 16:03:27.354Z"
}
```
Execution time: 4ms (Best performance observed)

On the Phone:
```json
{
  "collectionId": "wd3g9giw0wbb5tn",
  "collectionName": "pastes",
  "created": "2024-02-08 16:03:27.354Z",
  "id": "w0eakza33eaubjf",
  "text": "Hello",
  "updated": "2024-02-08 16:03:27.354Z"
}
```
Execution time: 14ms (Best performance observed)

Am I overlooking something, or are these results acceptable?

Initiall queries consistently take between 250-500ms to complete.

Additionally, when attempting to download a file on localhost, there's a latency of about 300-500ms before the request is processed. This increases to over 1s on hosting. For comparison, Discord processes the same file within 100ms.

Switching to PostgreSQL showed a dramatic improvement, with query times dropping to less than 1ms for a `SELECT *` query on a small database.

**Comment by ganigeorgiev** [comment]
If you are not able to reproduce it locally then I don't think your issue is PocketBase related.

The response times could vary for a lot of reasons (_location of the server, the available server resources, the client internet speed, disk speed, rate limiting applied by the host, etc._) and comparing the free fly.io teer to something like Discord doesn't make sense.

Accessing/Downloading image thumbs could be slow on first access because that is when the thumbs are being generated.

**Reply by ganigeorgiev** [reply]
Also your results doesn't match with the benchmarks - https://github.com/pocketbase/benchmarks (_see the [Fly.io test](https://github.com/pocketbase/benchmarks/blob/master/results/fly_free_tier.md), but again it would depend on a lot of factors_).

**Reply by ganigeorgiev** [reply]
Also:

> Switching to PostgreSQL showed a dramatic improvement, with query times dropping to less than 1ms for a SELECT * query on a small database.

Is PostgreSQL installed on the same free tier with the same parameters?
How you are even testing this?
Is the reported "1ms" response time the execution of a query that PostgreSQL reported or the time the query result was serialized and received by your client?

**Reply by ilshm** [reply]
Are response times of 4-14ms for localhost queries considered acceptable? However, experiencing a 500ms delay for file downloads on localhost seems excessively long.

**Reply by ilshm** [reply]
I meant localhost, with time less than 1ms on postgres (Windows machine)

**Reply by ganigeorgiev** [reply]
> Are response times of 4-14ms

Yes. This is a complete network request-response time, not just the query execution. There are queries that executes in nanosec. Start the executable with `--dev` and you'll see the reported query times.

**Reply by ganigeorgiev** [reply]
> I meant localhost, with time less than 1ms on postgres (Windows machine)

But what exactly is this "1ms"? If that 1ms is the reported of the PostgreSQL cli then ofcourse will be faster then a complete HTTP request.
If you run your query against `pb_data/data.db` with the sqlite3 cli tool I expect it to be even faster.

**Reply by ilshm** [reply]
Thank you for your help, but I'm still confused by the issues with file-related requests :(

**Reply by ilshm** [reply]
A 1ms response time actually exceeds the duration of the PostgreSQL query itself, as this includes waiting for the server's response via Express.js (as seen in Chrome DevTools), meaning the query execution time is even shorter. Haven't tried pocketbase for the same API yet, but expecting huge latency.

**Reply by ganigeorgiev** [reply]
> A 1ms response time actually exceeds ...

I doubt it as I'm not able to get consistently 1ms response time even for serving a static html with no db queries.

But I don't want to argue or convince you to anything. If you are satisfied with the PostgreSQL performance, just use PostgreSQL.

If you think that there is an issue with PocketBase, feel free to provide some reproducible steps and I'll try to investigate it but other than this I don't think I'll be able to help you further.

PocketBase performance is far from perfect but saying that it is "practically unusable" is pretty far fetched for me (_I'm using PocketBase on production with [presentator.io](https://presentator.io) and no one so far reported issues with the response times, nor I see any in the logs_).

**Reply by ilshm** [reply]
I totally get your perspective, and it wasn't my intention to portray PocketBase in a bad way. My comments were driven more by a desire to understand and optimize, rather than to criticize. Additionally, I'm relatively new to development, so there's a chance I might be missing something in my approach. I'll follow up later with steps to reproduce the issue. Thank you!

**Reply by saikoml** [reply]
@ganigeorgiev Shouldn't this be investigated?

**Reply by ganigeorgiev** [reply]
Investigate what exactly?

I don't even know what and how the author is performing their tests and as mentioned earlier performance (especially on the free Fly.io plan) could vary for a lot of reasons.

Unless there is a specific problem or bottleneck I don't know how to help further.

**Reply by ilshm** [reply]
@ganigeorgiev Apologies, it looks like the issue was on my end. After doing some tests and making changes to the projects, I'm now seeing query response times of about 0.5-1ms and the express api's are responding within 2-3ms ([Postgresql](https://github.com/ilyabumblebee/postgres-healthz) and [Pocketbase](https://github.com/ilyabumblebee/pocketbase-healthz)). No problems at all. Fly.io is generally reliable but occasionally slows down when handling the first request after a period of not being used. I'm still unsure about the earlier issues with the 15ms queries and the 500ms response times for files on localhost, as they seemed to resolve on their own, and I'm not sure what caused them. Thanks for your help, and sorry for any inconvenience.

**Reply by ganigeorgiev** [reply]
> Fly.io is generally reliable but occasionally slows down when handling the first request after a period of not being used. 

Make sure that `auto_stop_machines` is disabled and `min_machines_running` is set to 1 in your fly config (see https://fly.io/docs/apps/autostart-stop/#configure-automatic-start-and-stop).

Also note that in the above linked 2 `healthz` repos, the PocketBase example is doing a lot more work.

You are comparing raw PostgreSQL query against HTTP request to a full blown backend.
If you want a fair comparison against plain PostgreSQL you'll have to execute your query directly against the `pb_data/data.db` SQLite file (_or as minimum send direct client request to PocketBase without proxying the request through node_).

An oversimplification of the execution operations with your PostgreSQL example would be something like:
```
client request to node
    node route
        PostgreSQL query
          node response to client
```

With your PocketBase example the same will look something like:
```
client request to node
    node route
        node request to PocketBase
            PocketBase route
                SQLite query
                    PocketBase response to node
                        node response to client
```

To make the matter worst you are probably also not taking in consideration the fact that you are sending concurrently an admin `authWithPassword` request the first time when you initialize the SDK (_this is expected to be slow due to the bcrypt hash verification, not because of the database_).

And the above is just on the surface (even the executed queries wouldn't be similar because in the PocketBase example you have `sort`, aka. extra `ORDER BY`).

**PocketBase is not only a database and when you use its web APIs it doesn't return just the raw bytes of the queried table data.** It is a full blown backend application. When you send request to PocketBase we perform request parsing, data validations, authentication and access checks, response enriching/serialization/compression, etc.

**You can use your node orm with PostgreSQL, MySQL, SQLite, etc. directly. If you already have a node backend and just want a database then you may not need PocketBase.**

While PocketBase performance and memory consumption can definitely be improved, at the moment it is acceptable (at least for me) and further optimizations remains a low priority (unless there is a concrete issue/bottleneck).

**Reply by ilshm** [reply]
> To make the matter worst you are probably also not taking in consideration the fact that you are sending concurrently an admin authWithPassword request the first time when you initialize the SDK (this is expected to be slow due to the bcrypt hash verification, not because of the database).

If my understanding is correct, it authenticates once (at startup) and then utilizes this authentication for all subsequent API requests, which are made after the initial admin authentication.

> You can use your node orm with PostgreSQL, MySQL, SQLite, etc. directly. If you already have a node backend and just want a database then you may not need PocketBase.

Perhaps executing queries directly against a .db file could be an option? Additionally, leveraging PostgreSQL appears to be more challenging compared to PocketBase, especially in terms of user interface. Even self-hosting Supabase poses its own set of difficulties, can be slow on their AWS infrastructure, whereas PocketBase offers a user-friendly interface along with numerous built-in features.

P.S. Is there a method to ban or block users without the need to manually add and check ban columns?
